src/main/java/org/testng/TestRunner.java

```diff
@@ -961,7 +970,18 @@ public ISuite getSuite() {
 
   @Override
   public ITestNGMethod[] getAllTestMethods() {
-    return m_allTestMethods;
+    if (m_allTestMethods != null) {
+      //This is true only when we are running JUnit mode
+      return m_allTestMethods;
+    }
+    if (m_groupMethods != null) {
+      //Be it memory friendly mode or normal mode, we still have a
+      // configuration object that remembers all the test methods. So we always query it
+      return m_groupMethods.getAllTestMethods();
+    }
+    //If we are here, then it means that its ok to compute the methods every time
+    // Since we aren't being invoked from the core test execution code path.
+    return computeAndGetAllTestMethods();
```

Comment:
The last question is: if `m_groupMethods` is null here, why not creating one here with the `computeAndGetAllTestMethods`?
Or having another king of cache.

Reply:
It is going to be null only when a user turns on memory friendly mode and in that mode it has to be null.